src/node_native_module.h

```diff
@@ -23,93 +23,66 @@ using NativeModuleCacheMap =
-  // internalBinding('native_module').moduleIds
-  static void ModuleIdsGetter(v8::Local<v8::Name> property,
-                              const v8::PropertyCallbackInfo<v8::Value>& info);
-  // Passing config.gypi into JS land as internalBinding('native_module').config
-  static void ConfigStringGetter(
-      v8::Local<v8::Name> property,
-      const v8::PropertyCallbackInfo<v8::Value>& info);
-  // Get code cache for a specific native module
-  static void GetCodeCache(const v8::FunctionCallbackInfo<v8::Value>& args);
-  v8::MaybeLocal<v8::Uint8Array> GetCodeCache(v8::Isolate* isolate,
-                                              const char* id) const;
-  // Compile a specific native module as a function
-  static void CompileFunction(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static NativeModuleLoader* GetInstance();
 
   // Generated by tools/js2c.py as node_javascript.cc
   void LoadJavaScriptSource();  // Loads data into source_
   UnionBytes GetConfig();       // Return data for config.gypi
 
-  // Generated by tools/generate_code_cache.js as node_code_cache.cc when
-  // the build is configured with --code-cache-path=.... They are noops
-  // in node_code_cache_stub.cc
-  void LoadCodeCache();      // Loads data into code_cache_
-
-  // Compile a script as a NativeModule that can be loaded via
-  // NativeModule.p.require in JS land.
-  static v8::MaybeLocal<v8::Function> CompileAsModule(Environment* env,
-                                                      const char* id);
+  bool Exists(const char* id);
+  v8::Local<v8::Object> GetSourceObject(v8::Local<v8::Context> context);
+  v8::Local<v8::String> GetConfigString(v8::Isolate* isolate);
+  std::vector<std::string> GetModuleIds();
 
-  void InitializeModuleCategories();
   struct ModuleCategories {
     bool is_initialized = false;
     std::set<std::string> can_be_required;
     std::set<std::string> cannot_be_required;
   };
+  void InitializeModuleCategories();
+  const std::set<std::string>& GetCannotBeRequired();
+  const std::set<std::string>& GetCanBeRequired();
 
-  ModuleCategories module_categories_;
+  bool CanBeRequired(const char* id);
+  bool CannotBeRequired(const char* id);
+
+  NativeModuleCacheMap* code_cache();
+  v8::ScriptCompiler::CachedData* GetCodeCache(const char* id) const;
+  enum class Result { kWithCache, kWithoutCache };
+  // If an exception is encountered (e.g. source code contains
+  // syntax error), the returned value is empty.
+  v8::MaybeLocal<v8::Function> LookupAndCompile(
+      v8::Local<v8::Context> context,
+      const char* id,
+      std::vector<v8::Local<v8::String>>* parameters,
+      Result* result);
+  v8::MaybeLocal<v8::Function> CompileAsModule(v8::Local<v8::Context> context,
+                                               const char* id,
+                                               Result* result);
 
+  static std::unique_ptr<NativeModuleLoader> instance_;
```

Comment:
This probably should be an `std::optional` not a pointer.

Reply:
um, why should this be a `std::optional`? It is not optional, but a singleton?

Reply:
IIUC `std::optional` is the most similar monad to `Maybe`.
Actually why not keep it as a static member? It has a default constructor, and that way it will stay on the stack. Maybe later we can even turn into a `constexpr`.

Reply:
Why would this be a `Maybe`?

I think I can make it a static instance instead of a static pointer, I think I've done that in a previous iteration, but I don't remember why I turned this into a pointer anymore..